File: dp/ts/longest_palindromic_subsequence.ts

function longestPalindromeSubseq(s: string): number {
    
    const size: number = s.length;
    
    const dp: number[][] = new Array(size).fill(0).map(_ => new Array(size).fill(0));

    for (let i = 0; i < size;i++) {
        dp[i][i] = 1;
    }

    for (let i = size - 1; i >= 0; i--) {
        for (let j = i + 1; j < size; j++) {
            if (s[i] === s[j]) {
                dp[i][j] = dp[i + 1][j - 1] + 2;
            } else {
                dp[i][j] = Math.max(dp[i + 1][j], dp[i][j - 1]);
            }
        }
    }
    return dp[0][size - 1];
}

File: dp/golang/longest_palindromic_subsequence.go
package golang

// 53

// 516. 最长回文子序列
// https://leetcode.cn/problems/longest-palindromic-subsequence/description/

// 给你一个字符串 s ，找出其中最长的回文子序列，并返回该序列的长度。

// 子序列定义为：不改变剩余字符顺序的情况下，删除某些字符或者不删除任何字符形成的一个序列。

// 示例 1：

// 输入：s = "bbbab"
// 输出：4
// 解释：一个可能的最长回文子序列为 "bbbb" 。
// 示例 2：

// 输入：s = "cbbd"
// 输出：2
// 解释：一个可能的最长回文子序列为 "bb" 。

func longestPalindromeSubseq(s string) int {
	size := len(s)

	dp := make([][]int, size)

	for i := 0; i < size; i++ {
		dp[i] = make([]int, size)
		dp[i][i] = 1
	}

	for i := size; i >= 0; i-- {
		for j := i + 1; j < size; j++ {
			if s[i] == s[j] {
				dp[i][j] = dp[i+1][j-1] + 2
			} else {
				dp[i][j] = max(dp[i+1][j], dp[i][j-1])
			}
		}
	}

	return dp[0][size-1]
}
